### DISCUSSION TITLE
ESP32 and machine.reset()

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#9331] [Hardware & Peripherals] [answered]
https://docs.micropython.org/en/latest/library/machine.html
says ...

> Reset related functions
> 
> machine.reset()
> 
>     Resets the device in a manner similar to pushing the external RESET button.

but on an ESP32DevKitC dev board pressing the ENABLE button actually returns GPIO
to a low.  Whereas, machine.reset() will leave the GPIO (at least pin33) it its prior
state.

Does that doc also relate to the ESP32?

**Comment by robert-hh** [comment, marked as answer]
There is a difference between a hardware and a software action. `machine.reset()` does NOT change the voltage level at the EN pin, but causes the same code to be executed that pushing the EN pin. There can still be hardware related differences caused by wiring inside the chip, where pulling EN low may cause hardware state changes to some internal modules, which are not set like that by the software. That's why it's called `similar`.

Pulling the EN pin low at the ESP32 causes an internal power cycle, after which the code starts at a well defined address. The latter can be done by software as well. But the code may not reset internal states to the `power-on` state.

**Reply by davefes** [reply]
Thanks for the reply.  I now understand similar means ... not even close for my use-case!  I worked-around this by using a try/except like this:
```py
except KeyboardInterrupt as err:
    print ('You did a CTRL-C')
    print ('Do another CTRL-C within 5 seconds to go to REPL')
    print ('waiting 5 seconds')
    utime.sleep(5)

#  you didn't go to REPL so turn pin33 OFF before a machine.reset()
    pin33.off()
    machine.reset()
```
